### PR TITLE
Performance platform integration

### PR DESCRIPTION
**Pull request by dreamfall**

### CODE REVIEW

File: app/services/performance_platform_service.rb [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        data["_id"] = generate_applications_id(data)
+
+        result << data
+      end
+    end
+
+    result
+  end
+
+  private
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Useless `private` access modifier.

File: app/services/performance_platform_service.rb [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  #  }
+  #]
+
+  def self.perform_applications_by_stage
+    payload = fetch_applications_data
+
+    perform_request(APPLICATIONS_BY_STAGE_URL, payload)
+  end
+
+  def self.perform_request(url, payload)
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Assignment Branch Condition size for perform_request is too high. [16.52/15]

File: app/services/performance_platform_service.rb [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        http.request req
+      end
+
+      puts res.body
+
+      nil
+    end
+  end
+
+  def self.fetch_applications_data
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Assignment Branch Condition size for fetch_applications_data is too high. [35.74/15]

File: app/services/performance_platform_service.rb [hunk 4]
```diff
@@ -0,0 +1,197 @@
+require "json"
+require "net/http"
+require "base64"
+
+class PerformancePlatformService
+  TRANSACTIONS_BY_CHANNEL_URL = "https://www.performance.service.gov.uk/data/queens-awards-for-enterprise/transactions-by-channel"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Line is too long. [130/100]

File: app/services/performance_platform_service.rb [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  APPLICATIONS_BY_STAGE_URL = "https://www.performance.service.gov.uk/data/queens-awards-for-enterprise/applications-by-stage"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Line is too long. [126/100]

File: app/services/performance_platform_service.rb [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        result << data
+      end
+    end
+
+    result
+  end
+
+  private
+
+  # "A base64 encoded concatenation of: _timestamp, period, channel, channel_type, (the dimensions of the data point)"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Line is too long. [118/100]

File: app/services/performance_platform_service.rb [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    string = ""
+
+    %w(_timestamp period channel channel_type).each do |attr|
+      string << data[attr]
+    end
+
+    Base64.encode64(string)[0..7]
+  end
+
+  # "A base64 encoded concatenation of: _timestamp, period, award, stage, i.e. (the dimensions of the data point)"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Line is too long. [114/100]

File: app/services/performance_platform_service.rb [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+  end
+
+  def self.fetch_applications_data
+    result = []
+    timestamp = (Time.current - 1.week).beginning_of_day.utc
+
+    AWARD_TYPE_MAPPING.each do |award_type, award|
+      POSSIBLE_RANGES.each do |stage|
+        count = case stage
+        when "0-percent"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Indent `when` as deep as `case`.

File: app/services/performance_platform_service.rb [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+          form_answers
+            .where("fill_progress IS NULL OR fill_progress = 0")
+            .where(award_type: award_type)
+            .count
+        when "1-24-percent"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Indent `when` as deep as `case`.

File: app/services/performance_platform_service.rb [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+          form_answers
+            .where("fill_progress > 0 AND fill_progress < 25")
+            .where(award_type: award_type)
+            .count
+        when "25-49-percent"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Indent `when` as deep as `case`.

File: app/services/performance_platform_service.rb [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+          form_answers
+            .where("fill_progress >= 25 AND fill_progress < 50")
+            .where(award_type: award_type)
+            .count
+        when "50-74-percent"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Indent `when` as deep as `case`.

File: app/services/performance_platform_service.rb [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+          form_answers
+            .where("fill_progress >= 50 AND fill_progress < 75")
+            .where(award_type: award_type)
+            .count
+        when "75-99-percent"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Indent `when` as deep as `case`.

File: app/services/performance_platform_service.rb [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+          form_answers
+            .where("fill_progress >= 75 AND fill_progress < 100")
+            .where(award_type: award_type)
+            .count
+        when "100-percent"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Indent `when` as deep as `case`.

File: app/services/performance_platform_service.rb [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+  AWARD_TYPE_MAPPING = {
+    "trade" => "international-trade",
+    "innovation" => "innovation",
+    "development" => "sustainable-development",
+    "promotion" => "qaep"
+  }
+
+  POSSIBLE_RANGES = [
+   "0-percent",
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Use 2 spaces for indentation in an array, relative to the start of the line where the left bracket is.

File: app/services/performance_platform_service.rb [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+            .count
+        when "100-percent"
+          form_answers
+            .where("fill_progress = 100 OR submitted = true")
+            .where(award_type: award_type)
+            .count
+        end
+
+       data = {
+          "_timestamp" => timestamp.iso8601,
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.

File: app/services/performance_platform_service.rb [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+          "period" => "week",
+          "award" => award,
+          "stage" => stage,
+          "count" => count
+        }
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Indent the right brace the same as the start of the line where the left brace is.

File: app/services/performance_platform_service.rb [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+            .where(award_type: award_type)
+            .count
+        when "100-percent"
+          form_answers
+            .where("fill_progress = 100 OR submitted = true")
+            .where(award_type: award_type)
+            .count
+        end
+
+       data = {
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Inconsistent indentation detected.

File: app/services/performance_platform_service.rb [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+   "75-99-percent",
+   "100-percent"
+  ]
+
+  def self.run
+    perform_transactions_by_channel
+    perform_applications_by_stage
+  end
+
+  #[
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Missing space after #.

File: app/services/performance_platform_service.rb [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+  #  {
+  #      "_id": "23456780",
+  #      "_timestamp": "2015-03-10T00:00:00Z",
+  #      "period": "week",
+  #      "channel": "online",
+  #      "channel_type": "digital",
+  #      "count": 42
+  #  }
+  #]
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Missing space after #.

File: app/services/performance_platform_service.rb [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+      "count" => form_answers_count,
+      "_timestamp" => timestamp.iso8601
+    }
+
+    result["_id"] = generate_transactions_id(result)
+
+    perform_request(TRANSACTIONS_BY_CHANNEL_URL, [result])
+  end
+
+  #[
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Missing space after #.

File: app/services/performance_platform_service.rb [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+        data["_id"] = generate_applications_id(data)
+
+        result << data
+      end
+    end
+
+    result
+  end
+
+  # "A base64 encoded concatenation of: _timestamp, period, channel, channel_type, (the dimensions of the data point)"
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Line is too long. [118/100]

File: app/services/performance_platform_service.rb [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+  #  },
+  #  {
+  #      "_id": "23456780",
+  #      "_timestamp": "2015-03-10T00:00:00Z",
+  #      "period": "week",
+  #      "award": "qaep",
+  #      "stage": "0-percent",
+  #      "count": 42
+  #  }
+  #]
```
Comment (matthewford) [inline app/services/performance_platform_service.rb]: Missing space after #.

**Comment by RuslanHamidullin** [comment]
@dreamfall Looks good :+1: 
Can't merge - need to rebase

Some notes:

1) Looks like I need env var value for production servers:
```
ENV["PERFORMANCE_PLATFORM_TOKEN"]
```
Do we already have them?

2) In config/schedule.rb you probably can add locking
```
every :sunday, at: "12pm" do
  # with a small sleep say 3 secs before the work (in order to avoid duplicated running of jobs)
  runner "User.with_advisory_lock('platform_performance', 3) { PerformancePlatformService.run }"
end
```

**Comment by RuslanHamidullin** [comment]
@dreamfall so for your case would be:
```
every :sunday, at: "12pm" do
  runner "CronJobRun.run('performance', 0, 5) { PerformancePlatformService.run }"
end
```

**Comment by dreamfall** [comment]
@RuslanHamidullin thanks, will rebase and fix a bit later today.

**Comment by RuslanHamidullin** [comment]
@dreamfall ignore previous, would be:
```
every :sunday, at: "12pm" do
  runner "CronJobRun.run('performance') { PerformancePlatformService.run }"
end
```
=)

**Comment by matthewford** [comment]
@dreamfall @RuslanHamidullin this brings an interesting question - this should only be really run in prod - we need a switch in our schedule to say which tasks should run in which envs

**Comment by RuslanHamidullin** [comment]
got it - gonna do

**Comment by RuslanHamidullin** [comment]
:bomb:

**Pull request merged**